File: memsim/sim/fifo.py
from memsim.parser import parse_arguments, get_argument
from memsim.memory.base import send_request


class FIFO(object):
    """Class to simulate a FIFO between processes."""

    def __init__(self, index, size, item_size):
        """Create a simulated FIFO between processes.

        Arguments:
            index: A unique identifier for this FIFO.
            size: The size of the FIFO in items.
            item_size: The size of each item in the FIFO in bytes.
        """
        self.index = index
        self.size = size
        self.item_size = item_size
        self.offset = 0
        self.machine = None
        self.mem = None
        self.read_ptr = 0
        self.write_ptr = 0
        self.used = 0

    def __str__(self):
        result = '(fifo '
        result += '(id ' + str(self.index) + ')'
        result += '(size ' + str(self.size) + ')'
        result += '(item_size ' + str(self.item_size) + ')'
        result += ')'
        return result

    def total_size(self):
        return self.size * self.item_size

    def set_offset(self, offset):
        self.offset = offset

    def reset(self, machine, mem):
        self.machine = machine
        self.mem = mem
        self.mem.reset(machine)
        self.read_ptr = 0
        self.write_ptr = 0
        self.used = 0

    def is_full(self):
        return self.used == self.size

    def is_empty(self):
        return self.used == 0

    def done(self):
        return self.mem.done()

    def produce(self):
        """Put a value on the FIFO.

        Returns the access time or -1 if the FIFO is full.
        """
        if self.used == self.size:
            return -1
        else:
            addr = self.offset + self.write_ptr * self.item_size
            self.write_ptr = (self.write_ptr + 1) % self.size
            self.used += 1
            return send_request(self.mem, 0, True, addr, self.item_size)

    def consume(self):
        """Remove a value from the FIFO.

        Returns the access time or -1 if the FIFO is empty.
        """
        if self.used == 0:
            return -1
        else:
            addr = self.offset + self.read_ptr * self.item_size
            self.read_ptr = (self.read_ptr + 1) % self.size
            self.used -= 1
            return send_request(self.mem, 0, False, addr, self.item_size)


def parse_fifo(lexer):
    args = parse_arguments(lexer)
    index = get_argument(lexer, args, 'id', 0)
    size = get_argument(lexer, args, 'size', 1024)
    item_size = get_argument(lexer, args, 'item_size', 4)
    return FIFO(index=index, size=size, item_size=item_size)

File: fastsim/fifo.ml
open Machine
open Subsystem

class fifo =
    object (self)
        inherit subsystem as super

        val mutable min_depth : int = 1
        val mutable bram : bool = true
        val mutable read_ptr : int = 0
        val mutable write_ptr : int = 0
        val mutable used : int = 0

        method total_size = depth * word_size

        method set name value =
            match name with
            | "depth" ->
                depth <- max (int_of_string value) min_depth
            | "bram" ->
                bram <- value = "true"
            | "min_depth" ->
                min_depth <- int_of_string value;
                depth <- max depth min_depth
            | _ -> super#set name value

        method reset m main =
            super#reset m main;
            read_ptr <- 0;
            write_ptr <- 0;
            used <- 0;

        method is_full = used = depth

        method is_empty = used = 0

        method private process start write addr size =
            if bram || depth = 1 then start + 1
            else super#process start write addr size

        method produce =
            if used = depth then
                -1
            else
                let addr = write_ptr * word_size in
                begin
                    write_ptr <- (write_ptr + 1) mod depth;
                    used <- used + 1;
                    let result = self#process 0 true addr word_size in
                    score <- score + result;
                    result
                end

        method consume =
            if used == 0 then
                -1
            else
                begin
                    let addr = read_ptr * word_size in
                    let result = self#process 0 false addr word_size in
                    read_ptr <- (read_ptr + 1) mod depth;
                    used <- used - 1;
                    score <- score + result;
                    result
                end

        method peek offset =
            if used <= offset then
                -1
            else
                let temp = (read_ptr - offset) mod depth in
                let addr = temp * word_size in
                self#process 0 false addr word_size

    end
